File: apps/electron/src/mix.ts
/**
 * Offline mixdown of the two capture stems into one mono file.
 *
 * Streams the stems rather than loading them: an hour of 48 kHz audio is ~700 MB
 * as float32, and meetings run that long. Costs a second read of each stem,
 * because the peak has to be known before any sample can be scaled.
 */
import { open, type FileHandle } from "node:fs/promises";
import { WavWriter } from "./wav";

const BLOCK_FRAMES = 8192;

interface Format {
  readonly sampleRate: number;
  readonly channels: number;
  readonly bitsPerSample: number;
  readonly float: boolean;
  readonly dataStart: number;
  readonly dataEnd: number;
}

/** Walks the RIFF chunk list for `fmt ` and `data`. */
async function readFormat(handle: FileHandle): Promise<Format> {
  const head = Buffer.alloc(12);
  await handle.read(head, 0, 12, 0);
  if (head.toString("ascii", 0, 4) !== "RIFF" || head.toString("ascii", 8, 12) !== "WAVE") {
    throw new Error("not a RIFF/WAVE file");
  }

  let offset = 12;
  let format: Omit<Format, "dataStart" | "dataEnd"> | undefined;
  const header = Buffer.alloc(8);
  for (;;) {
    const { bytesRead } = await handle.read(header, 0, 8, offset);
    if (bytesRead < 8) break;
    const id = header.toString("ascii", 0, 4);
    const size = header.readUInt32LE(4);
    const body = offset + 8;

    if (id === "fmt ") {
      const chunk = Buffer.alloc(Math.min(size, 16));
      await handle.read(chunk, 0, chunk.byteLength, body);
      const tag = chunk.readUInt16LE(0);
      format = {
        channels: Math.max(1, chunk.readUInt16LE(2)),
        sampleRate: chunk.readUInt32LE(4),
        bitsPerSample: chunk.readUInt16LE(14),
        float: tag === 3,
      };
    } else if (id === "data") {
      if (!format) throw new Error("data chunk came before fmt");
      return { ...format, dataStart: body, dataEnd: body + size };
    }

    // chunks are word-aligned, odd sizes carry a pad byte
    offset = body + size + (size % 2);
  }
  throw new Error("no data chunk");
}

/**
 * One stem, read forward only. Holds just the frames the current block needs,
 * which is what keeps this a stream instead of a buffer.
 */
class Stem {
  readonly sampleRate: number;
  readonly frameCount: number;
  readonly #handle: FileHandle;
  readonly #format: Format;
  readonly #frameBytes: number;
  #cursor: number;
  #frames = new Float32Array(0);
  #base = 0;

  private constructor(handle: FileHandle, format: Format) {
    this.#handle = handle;
    this.#format = format;
    this.#frameBytes = (format.bitsPerSample / 8) * format.channels;
    this.#cursor = format.dataStart;
    this.sampleRate = format.sampleRate || 1;
    this.frameCount = Math.floor((format.dataEnd - format.dataStart) / this.#frameBytes);
  }

  static async open(path: string): Promise<Stem> {
    const handle = await open(path, "r");
    try {
      return new Stem(handle, await readFormat(handle));
    } catch (error) {
      await handle.close();
      throw error;
    }
  }

  get seconds(): number {
    return this.frameCount / this.sampleRate;
  }

  /** Buffer everything `at()` will ask for between `from` and `to` seconds. */
  async cover(from: number, to: number): Promise<void> {
    const first = Math.max(0, Math.floor(from * this.sampleRate));
    // +1 for the frame that interpolation reads ahead
    const last = Math.floor(to * this.sampleRate) + 1;
    if (first >= this.#base && last < this.#base + this.#frames.length) return;

    const keepFrom = Math.max(first, this.#base);
    const kept = this.#frames.subarray(Math.min(keepFrom - this.#base, this.#frames.length));
    const wanted = Math.max(last - keepFrom + 1, kept.length);
    // reads past the end of the stem stay zero, so a short stem fades to silence
    const frames = new Float32Array(wanted);
    frames.set(kept);

    let filled = kept.length;
    while (filled < wanted && this.#cursor < this.#format.dataEnd) {
      const bytes = Math.min(
        (wanted - filled) * this.#frameBytes,
        this.#format.dataEnd - this.#cursor,
      );
      const chunk = Buffer.allocUnsafe(bytes);
      const { bytesRead } = await this.#handle.read(chunk, 0, bytes, this.#cursor);
      if (bytesRead === 0) break;
      this.#cursor += bytesRead;
      filled += this.#decode(chunk.subarray(0, bytesRead), frames, filled);
    }

    this.#frames = frames;
    this.#base = keepFrom;
  }

  /** Interleaved samples to mono frames. Returns how many frames it wrote. */
  #decode(chunk: Buffer, into: Float32Array, at: number): number {
    const { channels, float, bitsPerSample } = this.#format;
    const width = bitsPerSample / 8;
    const frames = Math.floor(chunk.byteLength / this.#frameBytes);
    for (let frame = 0; frame < frames; frame++) {
      let sum = 0;
      for (let channel = 0; channel < channels; channel++) {
        const offset = frame * this.#frameBytes + channel * width;
        sum += float ? chunk.readFloatLE(offset) : chunk.readInt16LE(offset) / 32767;
      }
      into[at + frame] = sum / channels;
    }
    return frames;
  }

  /** The value at `seconds`, linearly interpolated. Silence outside the stem. */
  at(seconds: number): number {
    const position = seconds * this.sampleRate;
    const index = Math.floor(position);
    const current = this.#frame(index);
    return current + (this.#frame(index + 1) - current) * (position - index);
  }

  #frame(index: number): number {
    const offset = index - this.#base;
    return offset >= 0 && offset < this.#frames.length ? this.#frames[offset]! : 0;
  }

  async close(): Promise<void> {
    await this.#handle.close();
  }
}

async function eachBlock(
  systemPath: string,
  micPath: string | null,
  rate: number,
  frames: number,
  visit: (block: Float32Array) => Promise<void> | void,
): Promise<void> {
  const system = await Stem.open(systemPath);
  const microphone = micPath === null ? null : await Stem.open(micPath);
  const block = new Float32Array(BLOCK_FRAMES);

  try {
    for (let start = 0; start < frames; start += BLOCK_FRAMES) {
      const end = Math.min(start + BLOCK_FRAMES, frames);
      await system.cover(start / rate, (end - 1) / rate);
      await microphone?.cover(start / rate, (end - 1) / rate);

      for (let frame = start; frame < end; frame++) {
        const at = frame / rate;
        block[frame - start] = system.at(at) + (microphone?.at(at) ?? 0);
      }
      await visit(block.subarray(0, end - start));
    }
  } finally {
    await system.close();
    await microphone?.close();
  }
}

/**
 * Sums the stems into `outPath` at the system stem's rate, resampling the mic to
 * match. Returns the peak of the mix before any scaling.
 *
 * The two tracks run off independent clocks, so they are aligned at the start
 * and drift apart by whatever those clocks disagree on. Fine to listen back to,
 * not sample-accurate.
 */
export async function mix(
  systemPath: string,
  micPath: string | null,
  outPath: string,
): Promise<number> {
  const system = await Stem.open(systemPath);
  const rate = system.sampleRate;
  let seconds = system.seconds;
  await system.close();

  if (micPath !== null) {
    const microphone = await Stem.open(micPath);
    seconds = Math.max(seconds, microphone.seconds);
    await microphone.close();
  }
  const frames = Math.ceil(seconds * rate);

  let peak = 0;
  await eachBlock(systemPath, micPath, rate, frames, (block) => {
    for (const sample of block) {
      const magnitude = Math.abs(sample);
      if (magnitude > peak) peak = magnitude;
    }
  });

  // summing two loud tracks clips, so scale the mix back by exactly what it
  // overshot instead of clamping every sample
  const gain = peak > 1 ? 1 / peak : 1;
  const writer = await WavWriter.create(outPath, rate, 1);
  const scaled = new Float32Array(BLOCK_FRAMES);
  await eachBlock(systemPath, micPath, rate, frames, (block) => {
    for (let index = 0; index < block.length; index++) scaled[index] = block[index]! * gain;
    const view = scaled.subarray(0, block.length);
    writer.write(Buffer.from(view.buffer, view.byteOffset, view.byteLength));
  });
  await writer.finish();

  return peak;
}

File: apps/tauri/src-tauri/src/mix.rs
//! Offline mixdown of the two capture stems into one mono file.
//!
//! Streams the stems rather than loading them: an hour of 48 kHz audio is ~700
//! MB as f32, and meetings run that long. Costs a second read of each stem,
//! because the peak has to be known before any sample can be scaled.

use std::path::Path;

use hound::{SampleFormat, WavReader, WavSpec, WavWriter};

/// Sequential mono reader. Multi-channel stems are averaged down as they
/// arrive.
struct MonoReader {
    samples: Box<dyn Iterator<Item = Result<f32, hound::Error>>>,
    channels: usize,
}

impl MonoReader {
    fn open(path: &Path) -> Result<(Self, f64, u32), String> {
        let reader = WavReader::open(path).map_err(|error| error.to_string())?;
        let spec = reader.spec();
        let frames = reader.duration();
        let samples: Box<dyn Iterator<Item = Result<f32, hound::Error>>> = match spec.sample_format
        {
            SampleFormat::Int => {
                let scale = 1.0 / f32::from(i16::MAX);
                Box::new(
                    reader
                        .into_samples::<i16>()
                        .map(move |sample| sample.map(|value| f32::from(value) * scale)),
                )
            }
            SampleFormat::Float => Box::new(reader.into_samples::<f32>()),
        };
        let reader = Self {
            samples,
            channels: spec.channels.max(1) as usize,
        };
        Ok((reader, f64::from(spec.sample_rate), frames))
    }

    fn next_frame(&mut self) -> Option<f32> {
        let mut sum = 0.0;
        for channel in 0..self.channels {
            match self.samples.next() {
                Some(Ok(sample)) => sum += sample,
                // a torn frame at the tail is not worth failing the whole mix over
                Some(Err(_)) | None => return (channel > 0).then_some(sum / self.channels as f32),
            }
        }
        Some(sum / self.channels as f32)
    }
}

/// A stem being read at some other rate than its own. `at` only ever moves
/// forward, which is what lets this stay a stream instead of a buffer.
struct Track {
    source: MonoReader,
    rate: f64,
    /// Source index of `window[0]`.
    position: f64,
    window: [f32; 2],
}

impl Track {
    fn open(path: &Path) -> Result<(Self, f64, u32), String> {
        let (mut source, rate, frames) = MonoReader::open(path)?;
        let window = [
            source.next_frame().unwrap_or(0.0),
            source.next_frame().unwrap_or(0.0),
        ];
        let track = Self {
            source,
            rate,
            position: 0.0,
            window,
        };
        Ok((track, rate, frames))
    }

    /// The value at `seconds`, linearly interpolated. Reads past the end return
    /// silence, so a stem that stopped early just drops out of the mix.
    fn at(&mut self, seconds: f64) -> f32 {
        let target = seconds * self.rate;
        while target - self.position >= 1.0 {
            self.window[0] = self.window[1];
            self.window[1] = self.source.next_frame().unwrap_or(0.0);
            self.position += 1.0;
        }
        let fraction = (target - self.position).max(0.0) as f32;
        self.window[0] + (self.window[1] - self.window[0]) * fraction
    }
}

/// Frames the mix needs to cover both stems, at the output rate.
fn mixed_frames(stems: &[(f64, u32)], rate: f64) -> usize {
    stems
        .iter()
        .map(|(stem_rate, frames)| (f64::from(*frames) / stem_rate.max(1.0) * rate).ceil() as usize)
        .max()
        .unwrap_or(0)
}

fn each_frame(
    system_path: &Path,
    mic_path: Option<&Path>,
    frames: usize,
    rate: f64,
    mut visit: impl FnMut(f32) -> Result<(), String>,
) -> Result<(), String> {
    let (mut system, _, _) = Track::open(system_path)?;
    let mut microphone = match mic_path {
        Some(path) => Some(Track::open(path)?.0),
        None => None,
    };

    for frame in 0..frames {
        let at = frame as f64 / rate;
        let mut sample = system.at(at);
        if let Some(microphone) = microphone.as_mut() {
            sample += microphone.at(at);
        }
        visit(sample)?;
    }
    Ok(())
}

/// Sums the stems into `out` at the system stem's rate, resampling the mic to
/// match. Returns the peak of the mix before any scaling.
///
/// The two tracks run off independent clocks, so they are aligned at the start
/// and drift apart by whatever those clocks disagree on. Fine to listen back
/// to, not sample-accurate.
pub fn mix(system_path: &Path, mic_path: Option<&Path>, out: &Path) -> Result<f32, String> {
    let (_, system_rate, system_frames) = MonoReader::open(system_path)?;
    let rate = system_rate.max(1.0);

    let mut stems = vec![(system_rate, system_frames)];
    if let Some(path) = mic_path {
        let (_, mic_rate, mic_frames) = MonoReader::open(path)?;
        stems.push((mic_rate, mic_frames));
    }
    let frames = mixed_frames(&stems, rate);

    let mut peak = 0f32;
    each_frame(system_path, mic_path, frames, rate, |sample| {
        if sample.abs() > peak {
            peak = sample.abs();
        }
        Ok(())
    })?;

    // summing two loud tracks clips, so scale the mix back by exactly what it
    // overshot instead of clamping every sample
    let gain = if peak > 1.0 { 1.0 / peak } else { 1.0 };

    let spec = WavSpec {
        channels: 1,
        sample_rate: rate.round() as u32,
        bits_per_sample: 16,
        sample_format: SampleFormat::Int,
    };
    let mut writer = WavWriter::create(out, spec).map_err(|error| error.to_string())?;
    each_frame(system_path, mic_path, frames, rate, |sample| {
        let scaled = (sample * gain).clamp(-1.0, 1.0) * f32::from(i16::MAX);
        writer
            .write_sample(scaled.round() as i16)
            .map_err(|error| error.to_string())
    })?;
    writer.finalize().map_err(|error| error.to_string())?;

    Ok(peak)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::{env, f64::consts::TAU, fs, path::PathBuf};

    fn sine(name: &str, rate: u32, seconds: f64, hz: f64, amplitude: f32) -> PathBuf {
        let dir = env::temp_dir().join("mrec-mix-tests");
        fs::create_dir_all(&dir).unwrap();
        let path = dir.join(format!("{name}.wav"));
        let spec = WavSpec {
            channels: 1,
            sample_rate: rate,
            bits_per_sample: 16,
            sample_format: SampleFormat::Int,
        };
        let mut writer = WavWriter::create(&path, spec).unwrap();
        let frames = (f64::from(rate) * seconds).round() as usize;
        for frame in 0..frames {
            let value = (TAU * hz * frame as f64 / f64::from(rate)).sin() as f32 * amplitude;
            writer
                .write_sample((value * f32::from(i16::MAX)).round() as i16)
                .unwrap();
        }
        writer.finalize().unwrap();
        path
    }

    /// `(sample rate, channels, samples)` of a finished mix.
    fn read(path: &Path) -> (u32, u16, Vec<f32>) {
        let mut reader = WavReader::open(path).unwrap();
        let spec = reader.spec();
        let scale = 1.0 / f32::from(i16::MAX);
        let samples = reader
            .samples::<i16>()
            .map(|sample| f32::from(sample.unwrap()) * scale)
            .collect();
        (spec.sample_rate, spec.channels, samples)
    }

    fn loudest(samples: &[f32]) -> f32 {
        samples.iter().fold(0.0f32, |peak, s| peak.max(s.abs()))
    }

    #[test]
    fn resamples_the_mic_to_the_system_rate() {
        let system = sine("sys-rate", 48_000, 1.0, 440.0, 0.5);
        let mic = sine("mic-rate", 16_000, 1.0, 440.0, 0.5);
        let out = env::temp_dir().join("mrec-mix-tests/out-rate.wav");

        let peak = mix(&system, Some(&mic), &out).unwrap();
        let (rate, channels, samples) = read(&out);

        assert_eq!(rate, 48_000);
        assert_eq!(channels, 1);
        assert!(
            (samples.len() as i64 - 48_000).abs() <= 2,
            "{}",
            samples.len()
        );
        // two half-scale sines in phase land at full scale
        assert!((0.9..1.05).contains(&peak), "peak {peak}");
        assert!(loudest(&samples) <= 1.0001);
    }

    #[test]
    fn scales_back_instead_of_clipping() {
        let system = sine("sys-loud", 48_000, 0.5, 300.0, 0.9);
        let mic = sine("mic-loud", 48_000, 0.5, 300.0, 0.9);
        let out = env::temp_dir().join("mrec-mix-tests/out-loud.wav");

        let peak = mix(&system, Some(&mic), &out).unwrap();
        let max = loudest(&read(&out).2);

        assert!(peak > 1.7, "sum should overshoot, got {peak}");
        assert!(
            (0.99..=1.0001).contains(&max),
            "should normalise, got {max}"
        );
    }

    #[test]
    fn covers_the_longer_stem() {
        let system = sine("sys-short", 48_000, 0.2, 440.0, 0.4);
        let mic = sine("mic-long", 44_100, 1.0, 440.0, 0.4);
        let out = env::temp_dir().join("mrec-mix-tests/out-long.wav");

        mix(&system, Some(&mic), &out).unwrap();
        let (_, _, samples) = read(&out);

        assert!(
            (samples.len() as i64 - 48_000).abs() <= 2,
            "{}",
            samples.len()
        );
        // the mic keeps going after the system stem ends
        assert!(loudest(&samples[samples.len() - 1000..]) > 0.05);
    }

    #[test]
    fn passes_a_lone_system_stem_through() {
        let system = sine("sys-only", 48_000, 0.3, 1000.0, 0.6);
        let out = env::temp_dir().join("mrec-mix-tests/out-only.wav");

        let peak = mix(&system, None, &out).unwrap();
        let (rate, _, samples) = read(&out);

        assert_eq!(rate, 48_000);
        assert!(
            (samples.len() as i64 - 14_400).abs() <= 2,
            "{}",
            samples.len()
        );
        assert!((peak - 0.6).abs() < 0.01, "peak {peak}");
    }

    #[test]
    fn resampling_stays_continuous() {
        let system = sine("sys-seam", 48_000, 2.0, 220.0, 0.0);
        let mic = sine("mic-seam", 22_050, 2.0, 220.0, 0.3);
        let out = env::temp_dir().join("mrec-mix-tests/out-seam.wav");

        mix(&system, Some(&mic), &out).unwrap();
        let (_, _, samples) = read(&out);

        let jumps = samples
            .windows(2)
            .filter(|pair| (pair[1] - pair[0]).abs() > 0.05)
            .count();
        assert_eq!(jumps, 0, "{jumps} discontinuities in the resampled mic");
    }
}
